applications/serial_lte_modem/src/slm_at_httpc.c

```diff
@@ -0,0 +1,736 @@
+				return err;
+			}
+
+			if (at_params_valid_count_get(&at_param_list) == 5) {
+				err = at_params_int_get(&at_param_list, 4,
+							&httpc.sec_tag);
+				if (err < 0) {
+					LOG_ERR("Fail to get sec_tag: %d", err);
+					return err;
+				}
+				httpc.sec_transport = true;
+			}
+			LOG_DBG("Connect from http server");
+			if (do_http_connect() >= 0) {
+				err = 0;
+			} else {
+				err = -EINVAL;
+			}
+			break;
+		} else if (op == AT_HTTPCCON_DISCONNECT) {
+			LOG_DBG("Disconnect from http server");
+			err = do_http_disconnect();
+			if (err) {
+				LOG_ERR("Fail to disconnect. Error: %d", err);
+			}
+		} break;
+
+	case AT_CMD_TYPE_READ_COMMAND:
+		if (httpc.sec_transport) {
+			sprintf(rsp_buf, "#XHTTPCCON: %d,\"%s\",%d,%d\r\n",
+				(httpc.fd == INVALID_SOCKET)?0:1, httpc.host,
+				httpc.port, httpc.sec_tag);
+		} else {
+			sprintf(rsp_buf, "#XHTTPCCON: %d,\"%s\",%d\r\n",
+				(httpc.fd == INVALID_SOCKET)?0:1, httpc.host,
+				httpc.port);
+		}
+		rsp_send(rsp_buf, strlen(rsp_buf));
+		err = 0;
+		break;
+
+	case AT_CMD_TYPE_TEST_COMMAND:
+		sprintf(rsp_buf,
+			"#XHTTPCCON: (%d, %d),<host>,<port>,<sec_tag>\r\n",
+			AT_HTTPCCON_DISCONNECT, AT_HTTPCON_CONNECT);
+		rsp_send(rsp_buf, strlen(rsp_buf));
+		err = 0;
+		break;
+
+	default:
+		break;
+	}
+
+	return err;
+}
+
+
+/**@brief handle AT#XHTTPCREQ commands
+ *  AT#XHTTPCREQ=<method>,<resource>,<header>[,<payload_length>]
+ *  AT#XHTTPCREQ? READ command not supported
+ *  AT#XHTTPCREQ=?
+ */
+static int handle_AT_HTTPC_REQUEST(enum at_cmd_type cmd_type)
+{
+	int err = -EINVAL;
+	int param_count;
+	size_t method_sz = HTTPC_METHOD_LEN;
+	size_t resource_sz = HTTPC_RES_LEN;
+	size_t headers_sz = HTTPC_HEADER_LEN;
+	size_t offset;
+
+	if (httpc.fd == INVALID_SOCKET) {
+		LOG_ERR("Remote host is not connected.");
+		return err;
+	}
+
+	if (httpc.pl_len > 0) {
+		LOG_ERR("Another request is not finished.");
+		return err;
+	}
+
+	data_buf[HTTPC_BUF_LEN] = '\0';
+
+	switch (cmd_type) {
+	case AT_CMD_TYPE_SET_COMMAND:
+		param_count = at_params_valid_count_get(&at_param_list);
+		if (param_count < 3) {
+			return -EINVAL;
+		}
+		err = at_params_string_get(&at_param_list, 1,
+					   data_buf, &method_sz);
+		if (err < 0) {
+			LOG_ERR("Fail to get method string: %d", err);
+			return err;
+		}
+		data_buf[method_sz] = '\0';
+		httpc.method_str = data_buf;
+		offset = method_sz + 1;
+		/* Get resource path string */
+		err = at_params_string_get(&at_param_list, 2,
+					   data_buf + offset, &resource_sz);
+		if (err < 0) {
+			LOG_ERR("Fail to get resource string: %d", err);
+			return err;
+		}
+		data_buf[offset + resource_sz] = '\0';
+		httpc.resource = data_buf + offset;
+		offset = offset + resource_sz + 1;
+		/* Get header string */
+		err = at_params_string_get(&at_param_list, 3,
+					   data_buf + offset, &headers_sz);
+		if (err < 0) {
+			LOG_ERR("Fail to get option string: %d", err);
+			return err;
+		}
+		if ((offset + headers_sz + 1) > HTTPC_BUF_LEN) {
```

Comment:
Maybe we should just provide large-enough buffer to hold `HTTPC_METHOD_LEN + HTTPC_RES_LEN + HTTPC_HEADER_LEN + 3` by default?

Reply:
Now the buffer length(HTTPC_BUF_LEN) is set to the bigger one of HTTP request parameter or MTU
 - HTTPC_METHOD_LEN + HTTPC_RES_LEN + HTTPC_HEADER_LEN + 3
 - NET_IPV4_MTU

#define HTTPC_METHOD_LEN	20
#define HTTPC_RES_LEN		256
#define HTTPC_HEADER_LEN	512
#define HTTPC_REQ_LEN           (HTTPC_METHOD_LEN + HTTPC_RES_LEN \
				+ HTTPC_HEADER_LEN + 3)
#if HTTPC_REQ_LEN > NET_IPV4_MTU
#define HTTPC_BUF_LEN HTTPC_REQ_LEN
#else
#define HTTPC_BUF_LEN NET_IPV4_MTU
#endif